### PR TITLE
ETH fee-rate refactor 

### PR DESCRIPTION
**Pull request by jekabolt**
Improve ethereum fee-rate calculation according with MUL-2155

### CODE REVIEW

File: client/rest.go [hunk 1]
```diff
@@ -1348,25 +1349,9 @@ func (restClient *RestClient) getFeeRate() gin.HandlerFunc {
...
-			// 	restClient.log.Errorf("getFeeRate:currencies.Ether: no such networkid")
-			// }
-
-			if err != nil {
-				restClient.log.Errorf("getFeeRate:currencies.Ether:restClient.ETH.Cli: %v ", err.Error())
-			}
-			//speed, _ := strconv.Atoi(rate.GetGas())
 			switch networkid {
 			case currencies.ETHMain:
+				memppolTopFee := fetchMempool(restClient.ETH.Mempool)
```
Comment (nickpro) [inline client/rest.go]: let's correct double "p"

File: client/rest.go [hunk 2]
```diff
@@ -1460,15 +1445,29 @@ func (restClient *RestClient) getFeeRate() gin.HandlerFunc {
...
+		fee, ok := v.(int64)
+		if ok {
+			fees = append(fees, int64(fee))
+		}
+		return true
+	})
+	sort.Slice(fees, func(i, j int) bool { return fees[i] > fees[j] })
+	// take the average of n first feerate values to prevent ultra high rates
+	var quantile int64 = 5
+	fmt.Println(fees[:quantile])
```
Comment (nickpro) [inline client/rest.go]: please, remove prints in final code version

File: client/rest.go [hunk 3]
```diff
@@ -1460,15 +1445,29 @@ func (restClient *RestClient) getFeeRate() gin.HandlerFunc {
...
+	if int64(len(fees)) > quantile {
```
Comment (nickpro) [inline client/rest.go]: why dont we simply use ints here?

File: client/rest.go [hunk 4]
```diff
@@ -1460,15 +1445,29 @@ func (restClient *RestClient) getFeeRate() gin.HandlerFunc {
...
+		for _, fee := range fees[:quantile] {
+			maxFee += fee
+		}
+		return maxFee / quantile
 	}
-	return total / len(arr)
+
+	// standard very fast fee value if mempool is small
+	return int64(5 * 1000000000)
```
Comment (nickpro) [inline client/rest.go]: please, leave a commentary what is 5, what is 1000000000

File: client/rest.go [hunk 5]
```diff
@@ -1460,15 +1445,29 @@ func (restClient *RestClient) getFeeRate() gin.HandlerFunc {
...
 }
 
-func avg(arr []store.MempoolRecord) int {
-	total := 0
-	for _, value := range arr {
-		total += value.Category
-	}
-	if total == 0 {
-		return 0
+func fetchMempool(mempool sync.Map) int64 {
```
Comment (nickpro) [inline client/rest.go]: looks like we can cover this func with unittest

File: client/rest.go (hunk 3 above)
Comment (jekabolt) [reply client/rest.go]: Gas price can overflow int32

File: client/rest.go (hunk 5 above)
Comment (jekabolt) [reply client/rest.go]: ok

File: client/rest.go [hunk 6]
```diff
@@ -1348,25 +1348,11 @@ func (restClient *RestClient) getFeeRate() gin.HandlerFunc {
...
-
-			if err != nil {
-				restClient.log.Errorf("getFeeRate:currencies.Ether:restClient.ETH.Cli: %v ", err.Error())
-			}
-			//speed, _ := strconv.Atoi(rate.GetGas())
+			// first "quantile" sotrted in descending order of mempool for definition of highest rate
+			var quantile int = 5
 			switch networkid {
 			case currencies.ETHMain:
+				mempolTopFee := fetchMempool(restClient.ETH.Mempool, quantile)
```
Comment (nickpro) [inline client/rest.go]: so close
double O :D

**Review by nickpro — APPROVED** [review]

**Pull request merged**